rfcs/0000-rlp-cleanup.md

```diff
@@ -0,0 +1,331 @@
+
+### Goals of cleaning up the RLP
+
+1. Decouple the API from the underlying implementation - i.e. provide a more generic abstraction
+2. Simplify the API - i.e. DRY, straight to the point (for defining rate limits)
+3. Consistency with Kuadrant's [AuthPolicy](https://pkg.go.dev/github.com/kuadrant/kuadrant-operator@v0.2.1/api/v1beta1#AuthPolicy) API - i.e. same language, similar UX
+
+### Current WIP to consider while discussing the cleaning up of the RLP
+
+1. Inheritance of the network matching rules from the upstream network object (https://github.com/Kuadrant/architecture/pull/4)
+2. Improvements to RateLimitPolicy - aka "minimal working example" without `configurations`, `limits.conditions` and `limits.variables` (https://github.com/Kuadrant/kuadrant-operator/issues/133)
+3. Implement `skip_if_absent` for the RequestHeaders action (https://github.com/Kuadrant/wasm-shim/issues/29)
+4. Rate Limiting across clusters ([doc](https://docs.google.com/document/d/1pqCODRAkNUTLB6lJfRWcv3d2Hlj9WqsGySmjrP707Vk/edit#heading=h.nzpgr3pef6uy))
+
+## Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Given the following HTTPRoute:
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1alpha2
+kind: HTTPRoute
+metadata:
+  name: toystore
+  labels:
+    app: toystore
+spec:
+  parentRefs:
+    - name: istio-ingressgateway
+      namespace: istio-system
+  hostnames: ["*.toystore.com"]
+  rules:
+    - matches:
+        - path:
+            type: PathPrefix
+            value: "/toys"
+          method: GET
+        - path:
+            type: PathPrefix
+            value: "/toys"
+          method: POST
+      backendRefs:
+        - name: toystore
+          port: 80
+```
+
+These are examples of RateLimitPolicies targeting the route:
+
+**Example 1.** Minimal example - unconditional, unqualified rate limiting
+
+```yaml
+apiVersion: kuadrant.io/v2beta1
+kind: RateLimitPolicy
+metadata:
+  name: toystore
+spec:
+  targetRef:
+    group: gateway.networking.k8s.io
+    kind: HTTPRoute
+    name: toystore
+  limits:
+    - rates:
+        - limit: 5
+          unit: minute
+```
+
+**Example 2.** All you can get - conditions, counter qualifiers, multiple rates, and weights
+
+```yaml
+apiVersion: kuadrant.io/v2beta1
+kind: RateLimitPolicy
+metadata:
+  name: toystore
+spec:
+  targetRef:
+    group: gateway.networking.k8s.io
+    kind: HTTPRoute
+    name: toystore
+  limits:
+    - name: writers
+      selectors:
+        # this is a condition
+        - selector: request.http.method
+          operator: eq
+          value: POST
+        # this is counter qualifier
+        - selector: metadata.auth.identity.username
+      rates:
+        - limit: 5
+          duration: 1
+          unit: minute
+        - limit: 100
+          duration: 12
+          unit: hour
+      weight: 1
```

Comment:
I am unfamiliar with the weight concept when talking about rate limting?

Reply:
@maleck13, it could also be called "increment by".

This is a concept in parts inherited from [3scale](https://access.redhat.com/documentation/de-de/red_hat_3scale_api_management/2.7/html-single/getting_started/index#defining-mapping-rules) and, at the same time, a placeholder to kick off the discussion around a more sophisticated rate limit system where the increments are not necessarily linear to the number of hits. I'm thinking ways to increment the counter based upon variable factors such as the size of the payload, the estimated computation effort to process the request, etc – to name a few.

At the very least, in its simplest form, the option to modify the weight (or increment) of a limit can be a way for users to play with the rates without resetting the counter. It opens up implementation options as well for what's sometimes called "dynamic rate limit".

In time: I now wonder if "increment" isn't indeed a better name for the thing, so to avoid the confusion with something else, like weighting between limits (e.g. to infer priority).

Reply:
Yeah increment seems more explicit. I def think this would be an optional field with a default of 1 though

Reply:
what about **delta** optional field which defaults to 1?

Reply:
I'm unsure whether we should add `delta` at this stage, but won't fight it neither… otoh, I wonder if a `disabled` or something flag could be useful? i.e. _not_ using a magic value as a limit (e.g. `0`)

Reply:
Removed from the RFC.

As agreed, the increment feature is out of scope. It deserves its own proposal to be discussed separately.